### ISSUE TITLE
[Custom Elements] the `extends` property cannot be replaced

### ISSUE THREAD

**Issue by WebReflection** [WICG/webcomponents#326]
.

**Comment by WebReflection** [comment]
I've been  told to CC @annevk , @rniwa , @hayatoito , and @travisleithead about this, so here i am.

The long discussion about `is` property and the `extends` one is that "anyone could wrap an input, a select, a form ... " etc, etc ...

When it comes to partial applications, not every time the element can be "just inherited", and the `is` behavior, as well as the `extends` one, is the only way to resolve the issue.

### Example

In a partial application we might want to create a `CustomElement` that should act like a `HTMLTableRowElement`

``` js
var MyTr = document.registerElement('my-tr', {
  prototype: Object.create(HTMLTableRowElement.prototype)
});
var mtr = new MyTr;
mtr.innerHTML = '<td>1</td><td>2</td>';
mtr.innerHTML // "12" <== SEE THAT? EPIC FAIL
```

If the component just inherits `HTMLTableRowElement.prototype` it can be used to be appended to a generic `tbody` but it won't preserve its content as html.

Only using the good old way works:

``` js
var mtr = document.createElement('tr');
mtr.innerHTML = '<td>1</td><td>2</td>';
mtr.innerHTML // "<td>1</td><td>2</td>" <== OK
```

Partial rows,  cells, cols or any other table related application can be very useful.

As example, [this replication of the (in)famous DBMonster benchmark](https://github.com/WebReflection/dom-class/blob/master/demo/db-monster.js) uses 3 components:
- the `HTMLTableElement` one to create a table component
- the `HTMLTableRowElement` one to create [each row](https://github.com/WebReflection/dom-class/blob/master/demo/db-monster.js#L86) of such table
- the `HTMLTableCellElement` to create [last 5 cells](https://github.com/WebReflection/dom-class/blob/master/demo/db-monster.js#L33) of each rows

The [benchmark page](https://github.com/WebReflection/dom-class/blob/master/demo/dbmonster.html) as it is [works like a charm](http://webreflection.github.io/dom-class/demo/dbmonster.html) in every [polyfilled](https://github.com/WebReflection/document-register-element) platform but it would be impossible to recreate such simple use case (even if stressed in this bench) without the usage of the `is` specification.

### How it's currently solved

Well, the components that inherits `TableRow` and `TableCell` are created in this way:

``` js
var MyTr = document.registerElement('my-tr', {
  extends: 'tr',
  prototype: Object.create(HTMLTableRowElement.prototype)
});
mtr.innerHTML = '<td>1</td><td>2</td>';
mtr.innerHTML // "<td>1</td><td>2</td>" <== Now We Talk!
```

So that only after that it's possible to assign `td` related content and preserve it.

The resulting `<tr is="my-tr"></tr>` is at the end of the day **more than welcome** and it never fails in  both modern and polyfilled browsers.

Thanks for your consideration, I should also probably mention that this is not necessarily a bug in the specs in my opinion,  but rather a point to let `is` and `extends` survive the current specifications: it would break badly in some  case otherwise and with not evergreen/legacy browsers.

Best Regards

**Comment by WebReflection** [comment]
I'd like to add more content/reasons/examples since I've had this conversation privately via email too.

I once demoed [an image with an src pointing to a static map](http://webreflection.github.io/document-register-element/test/examples/x-map.html).

That image was an `<img is="x-map">`, it was working on every browser available in planet earth by default, then for those browsers polyfilled via my [document-register-element](https://github.com/WebReflection/document-register-element), including IE8, the image was becoming a LeafLet map.

How would you style or search for an element that is a custom one?

In the image case `img[is=x-map], x-map` did a great job!

Was the image one of those cases where having `<x-map><img></x-map>` could have worked regardless?

Technically yes, practically why would we like an extra layer we don't even know how old quirky browsers could render by default (block? padding?)?
Why having indirection when registering 'x-map' as `'img'` extends is all it takes to have also a cleaner layout?

This img case without `extends` would be similarly cumbersome for `input`, `select`, `textarea`, `button` and all form elements and other nodes.

Do we really need an overblown DOM tree when we can simply register through the `extends` and address them  via `[is=x-thing]` ?

I think this would also be a very straight forward simplification:
- we want our own full new thing? `prototype: HTMLNativeElement`
- we want the previous one plus a native like appearance and behavior? `extends: 'native'`

Otherwise we need to also add listeners that inevitably have to bubble up, 'cause if we have an input component that doesn't directly trigger `input` events but only delegates, it also has to delegate back as well to the input, right?

But why would we like to increase complexity to create Custom Elements that could just work?

Back to my initial TR example, avoiding the usage of `innerHTML`, since that is not the real problem, check this example:

``` html
<!DOCTYPE html>
<html>
  <head>
    <script>
    var XTR = document.registerElement('x-tr', {
      prototype: Object.create(HTMLTableRowElement.prototype)
    });
    </script>
  </head>
  <body>
    <table><x-tr><td>hello</td></x-tr></table>
  </body>
</html>
```

If above code would be everything I need in order to create tables with super-powered rows, there won't be any problem.

Unfortunately if we **inspect the page layout** ... this is what we are going to find there:

``` html
<!DOCTYPE html>
<html>
  <head>
    <script>
    var XTR = document.registerElement('x-tr', {
      prototype: Object.create(HTMLTableRowElement.prototype)
    });
    </script>
  </head>
  <body>
    <x-tr></x-tr> <!-- SEE THAT? THAT'S NOT WHAT WE WROTE! -->
    <table><tbody><tr><td>hello</td></tr></tbody></table>
</body></html>
```

In other words, the mechanism provided to create Custom Element completely destroyed a layout and its meant behavior. The app is doomed and nothing is even warning us about it.

Accordingly, the only way to make it work is using the alternative "_need native-like behavior_" formula based on `extends`. Following the tiny change full of win!

``` html
<!DOCTYPE html>
<html>
  <head>
    <script>
    var XTR = document.registerElement('x-tr', {
      extends: 'tr',
      prototype: Object.create(HTMLTableRowElement.prototype)
    });
    </script>
  </head>
  <body>
    <table><tr is="x-tr"><td>hello</td></tr></table>
  </body>
</html>
```

Now we can use that row in every table we want, it will be compatible with years of styles and legacy based on tr and for modern browsers it will act like a super powered row.

How cool is that? And why shouldn't Custom Elements be this cool and powerful? Why would we like to limit them  instead of simply agreeing that if it's native like that's needed, then we need to use `extends` and if we want to be sure we select an expected component, `[is="x-name"],xname` is all we need to have it back either ways?

As summary, this is the reason I've filed this bug: AFAIK the `is` attribute is in danger, but I believe we are not really thinking through how many things will become impossible to achieve ... and I wonder for what counter-achievement ... one less attribute to check beside `class`, `id`, `style`, and others? Is that what is all this about or there is some problem I am not seeing? (if so please tell me more on `extends` and `is` side effects, thank you)

I hope now it's even cleaner what is the problem here, and why `extends` is the ideal solution.

Best Regards

**Comment by bkardell** [comment]
@WebReflection there's a lot here, it's tough to respond to so much in a concise way so I'll try to ask questions/make comments in a sensible manner.

To the best of my understanding, it was agreed to postpone form of extension to v2 because the issues raised were pretty myriad and after literally years of discussion there wasn't headway on getting agreement that enough of them were solved.  On the other hand, there was much closer agreement on things that _could_ get into v1 within a reasonable time frame and move efforts forward making a whole bunch of new things possible.  There are also corresponding efforts in other areas which potentially change the inputs to the discussion about how to solve bits of it.  I honestly don't know that anyone has or is willing to take a really hard dive into _all_ of the notes from emails/ml threads/discourse/tweets/etc and document them - I know a few efforts were made, but they all seem incomplete to me.  In any case, the point is that when you say it's "in danger" it's not like everyone has given up on it completely.  Note also that anything that you've done to-date could still effectively continue to work with minimal tweak because you have an emulation that you feel is reasonable.

More specifically to your points/examples - they seem actually quite different to me.. It's hard to discuss this without knowing what you would do if you had shadow dom on your side, but I'll try:  In the leaflet map example you say "this image is a map" - but it's actually only briefly an image, the image is actually destroyed and replaced with an x-map which is no longer an image.  ATs don't understand it as an image, but it will be understood in terms of the x-map which is a composed thing full of native elements.   Further, the APIs for images don't automatically make sense (how would this deal with alt, sizes, src-set, how would CSS deal with heights, etc, the events don't map and so on - images don't have children but x-map does...plus you have to understand both states for querying and styling and so on).  So what you've solved nicely there is progressive enhancement I think - but the wrapper could do that as well without the distortions of claiming to be something it isn't - instead it could decorate input or fallback.  All of these (I think) represent _some_ of the common questions/solutions raised with is on most elements (especially some of the most common ones like buttons and inputs).

If I'm being honest, I don't think this is as strong a case as your tables one - or infact - anything which requires a special context.  Here's why that one doesn't particularly bother me so much though - If you put a `<td>foo</td>` (or several other tags) in the body of a page, it'll parse to plain text.  If you create one with createelement you can append it in weird places sometimes, but it won't work.  In some cases, fixups happen, in some cases, not.  The same limitations hold/affect on innerHTML, document fragments and so on.   These elements are weird and without some way to tap into the parser it seems to me they'd always have the kind of problems like above (that's not something special to custom elements, it applies to regular html as well).  FWIW though, in your final example it does seem to me you could use the wrapper approach there as well with a little more indirection - effectively decorate to bestow upon a standard table that any standard row becomes a super-row.  Yes, it's imperfect, but it's imperfect in a not-actually-different-than-the-existing-platform kind of way that (I think) everyone hopes to solve better in the future.

Still, if there were agreement that this is especially a problem for contextual element - I wonder if it is possible to consider some way to solve this much more focused problem in a simpler way fairly early?

**Comment by WebReflection** [comment]
Hi @bkardell , I know it has been a long discussion, yet you didn't tell me _why is that a problem_ concretely, this part I miss.

You've also commented the "extra content" second example more than the main issue I've initially reported: table rows, cells, captions.

I wrote it's not as compelling use case as table rows or cells because a wrap might just do, yet I dont understand any of your points. Methods? If it's not recognized as Custom Element due browser limitations there's nothing to worry about, it's an image.

`width` and `height` ? Already used and styled without problems. You are basically saying there are issues in something that has no issue at all and already works in prod even on IE8.

Regardless, once again, that is not the main issue: LI, TR, TD, COL, CAPTION, other elements outside their parent context _are_ the issue so let's focus on these cases ;-)

If we have "some element can be extended, some other can't" the API will fail in consistency, and developer expectations will be compromised like shown in the `x-tr` example.

> in your final example it does seem to me you could use the wrapper approach 

Please provide an example of custom TR that works in a regular table (simply because of its peculiar layout that works like that since the beginning of the time), otherwise I don't see how any wrapper could work.

The only way I see this solvable without `extend` and `is` is improving specifications in a way that if the inherited object is "one of those special cases" then it should be accepted as exception whenever it's supposed to be accepted (a `tbody` element in this case)  without ever loosing it's ability to be populated via `innerHTML` like its inherited counterpart would do.

Let's remember that this works already since ever:

``` js
var tr  = document.createElement('tr');
tr.innerHTML = '<td>yes, it works as easy as this</td>';
```

And let's think  on  why on earth something  that extends a TR shouldn't work exactly same way.

Thanks

**Comment by jarek-foksa** [comment]
Here is another use case.

I'm using is-attributes in my [SVG editor](https://chrome.google.com/webstore/detail/boxy-svg/gaoogdonmngmdlbinmiclicjpbjhgomg) to implement custom shapes such as a star, a ring or an arrow by extending the `<path>` element. For example when user draws a star, I'm generating the following markup using my custom [star shape API](https://gist.github.com/jarek-foksa/151418a0b4ab6ebab330):

``` svg
<path is="bx-starpath" d="..." data-bx-cx="204" data-bx-cy="153" data-bx-r1="41" data-bx-r2="103" data-bx-arms="7"></path>
```

You might argue that I could achieve the same result with <bx-star> custom element that wraps the path:

``` svg
<bx-star cx="204" cy="153" r1="41" r2="103" arms="7">
  <-- Shadow DOM -->
  <path d="..."></path>
</bx-star>
```

The problem with the latter approach is that the custom element will render correctly only in environment that supports scripting and includes my implementation of the star shape, it won't work when loaded into another SVG editor such as Adobe Illustrator or when set as a background image.

**Comment by domenic** [comment]
It's not really clear what the bug with the spec is that this is about. Closing, but happy to reopen if there's a concrete bug here.
